### PR TITLE
Record All2All split sizes so torchfrtrace can name the mismatched peer pair

### LINKED ISSUE #194559: [c10d][Flight Recorder] Record All2All split sizes so torchfrtrace can name the mismatched peer pair
### 🚀 The feature, motivation and pitch

Currently `torchfrtrace` cannot tell which ranks broke an All2All ; per-rank sizes are allowed to differ, so the analyzer marks it `UNDECIDED` and only checks global numel, as seen here: 

https://github.com/pytorch/pytorch/blob/main/torch/distributed/flight_recorder/components/utils.py#L416-L422

```python
def check_size_alltoall(alltoall_cases: list[dict[str, Any]]) -> tuple[bool, int, int]:
    input_numel = 0
    output_numel = 0
    for e in alltoall_cases:
        input_numel += math.prod(e["input_sizes"][0])
        output_numel += math.prod(e["output_sizes"][0])
    return input_numel != output_numel, input_numel, output_numel
```

So when it fails ,I  just get two totals, as picking a culprit rank is obviously hard with this methodology. However, instead I think we can rely on edges of  rank(i,j) where the logic follows: 

```text
elements rank i sends to j  ==  elements rank j receives from i
```

Such that on mismatch, we print that pair and both stacks, e.g.

```text
All2All split mismatch, collective sequence number: 1
  rank 0 send -> rank 1: 3 elements
  rank 1 recv <- rank 0: 1 elements
```

To me it  is more useful on an NCCL timeout, than global numbers.

***Proposal*** (All2All only)

- Record optional `input_split_sizes` / `output_split_sizes` (element counts, dest/src indexed) on the FR entry. Attach after `record()`.
- For `torchfrtrace`, pairwise edge check when splits (or even/list reconstruction) exist; else we keep the global check.

Have a working solution, happy to send a PR if this direction is ok.

### Alternatives

_No response_

### Additional context

_No response_

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @kapilsh

### PR DESCRIPTION
**Pull request by substobeme**
Fixes #194559:

For Capture: 
I did not want to grow `record()` (every collective goes through that). So I attach the lists afterwards with `record_split_sizes`. Where that happens depends on the backend:

- Native NCCL / Gloo: after the collective is issued
- Hook backends (`nccl2` / `fake`): the pre-hook now carries the dim-0 lists for `all_to_all_single`; list `all_to_all` I just take `numel` per tensor (one tensor per rank already)

`all_to_all_single` splits are dim-0 row counts, so I multiply by `numel / size(0)` and dump **element counts**.  

Schema bump is minor so updated version from`2.10`to `2.11`,( both JSON and pickle write them).

Old dumps without the keys still fall back to the global numel check, so I think we stay backwards compatible.

For Analyzer:
Pairwise check only runs after `UNDECIDED` (that is all All2All). If the split lists are missing or the wrong length, I fall back to `check_size_alltoall`, and I fixed that to sum **all** tensors instead of just `input_sizes[0]`.
I also made sure  not to invent an even split when the keys are absent. That was false-positiving legal uneven All2All on pre-2.11 dumps (rank 0 `[[8]]` / rank 1 `[[4]]` with matching totals per rank, no keys).

Added tests for:
- analyzer pairwise edges (healthy, broken, broken even when global totals agree)
- 2.10 dumps with no split keys (global numel fallback, and the even-split false positive on legal uneven All2All)
- live `all_to_all_single` dump on NCCL, Gloo, and the hook (`fake` / `nccl2`) ; element counts on pickle and JSON

Repro for NCCL(script.py) for 2 GPU setup:

```python
import os
import pickle
from pathlib import Path

import torch
import torch.distributed as dist

local_rank = int(os.environ["LOCAL_RANK"])
dev = local_rank if torch.cuda.device_count() > 1 else 0
torch.cuda.set_device(dev)
dist.init_process_group(
    "nccl",
    device_id=torch.device("cuda", dev),
)
rank = dist.get_rank()

inp = torch.ones(8, device="cuda")
if rank == 0:
    out = torch.empty(8, device="cuda")
    dist.all_to_all_single(
        out, inp, output_split_sizes=[5, 3], input_split_sizes=[5, 3], async_op=True
    )
else:
    out = torch.empty(4, device="cuda")
    dist.all_to_all_single(
        out, inp, output_split_sizes=[1, 3], input_split_sizes=[5, 3], async_op=True
    )

dump = pickle.loads(
    torch._C._distributed_c10d._dump_nccl_trace(includeStackTraces=False)
)
for e in dump.get("entries") or []:
    e["state"] = "scheduled"

prefix = os.environ["TORCH_FR_DUMP_TEMP_FILE"]
Path(prefix + str(rank)).write_bytes(pickle.dumps(dump))
os._exit(0)
```
Bash:
``` bash
mkdir -p /tmp/fr_nccl
TORCH_FR_BUFFER_SIZE=10 TORCH_FR_DUMP_TEMP_FILE=/tmp/fr_nccl/rank \
  torchrun --nproc_per_node=2 script.py
python -m torch.distributed.flight_recorder.fr_trace /tmp/fr_nccl --prefix rank -v
```

Output(Important part):

``` loaded 2 files in 0.0001480579376220703s
built groups, memberships
All2All split mismatch, collective sequence number: 1
internal record id: 0
group info: 0:default_pg
collective: nccl:all_to_all
input sizes: [[8]]
output sizes: [[8]]
world size: 2
expected ranks: {0, 1}
collective state: scheduled
  rank 0 send -> rank 1: 3 elements
  rank 1 recv <- rank 0: 1 elements
  rank 1 send -> rank 0: 5 elements
  rank 0 recv <- rank 1: 3 elements

```

**Note:** Still in draft since in the PR split index is group-local rank, but the analyzer currently maps peers with `sorted(global ranks)`, same as P2P `_init_global_src_dst`. That is wrong if someone did `new_group(ranks=[4, 0, 2])` (not already ascending). Default WORLD and the usual `[0, 2, 4]` subgroups are fine, which is most of the NCCL timeout cases I care about, nevertheless I want to look into this edge case.

### CODE REVIEW

File: torch/csrc/distributed/c10d/ProcessGroupGloo.cpp [hunk 1]
```diff
@@ -2448,6 +2448,10 @@ c10::intrusive_ptr<Work> ProcessGroupGloo::all_to_all_single(
   c10d::checkSplitSizes(inputCounts, inputTensor, getSize());
   c10d::checkSplitSizes(outputCounts, outputTensor, getSize());
 
+  auto in_splits =
```
Comment (thisisatharva-rh) [inline torch/csrc/distributed/c10d/ProcessGroupGloo.cpp:2451]: I would move the gloo changes to a seperate PR

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -5420,6 +5420,13 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::all_to_all_single(
         OpType::ALLTOALL_BASE,
         opts.asyncOp,
         "nccl:all_to_all");
+    auto* w = static_cast<WorkNCCL*>(work.get());
```
Comment (thisisatharva-rh) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:5423]: i am skeptical about this, when asyncOP=false in all2all, wouldn't collective() return a nullptr? could cause segfaults

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 3]
```diff
@@ -5474,6 +5481,14 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::all_to_all_single(
...
```
Comment (thisisatharva-rh) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:5484]: same here

File: torch/csrc/distributed/c10d/FlightRecorderDetail.hpp [hunk 4]
```diff
@@ -269,6 +269,23 @@ std::optional<typename FlightRecorder<EventType>::Entry> FlightRecorder<
...
+    std::optional<size_t> id,
+    std::optional<size_t> reset_epoch,
+    std::vector<int64_t> input_split_sizes,
+    std::vector<int64_t> output_split_sizes) {
+  if (!enabled_ || !id || !reset_epoch) {
+    return;
+  }
+  std::lock_guard<std::mutex> guard(mutex_);
+  Entry* entry = &entries_.at(getIdxFromId(*id, *reset_epoch));
+  if (entry->id_ == *id && entry->reset_epoch_ == *reset_epoch) {
```
Comment (thisisatharva-rh) [inline torch/csrc/distributed/c10d/FlightRecorderDetail.hpp:283]: if this entry got recycled by the time we patch the splits, we lose the split sizes. can we set this in record()

**Review by thisisatharva-rh — CHANGES_REQUESTED** [review]
Changes required, especially on the nullptr dereferences
